### PR TITLE
Refactor validation: introduce the Consensus trait

### PR DESCRIPTION
**Pull request by aakoshh**
**Summary of changes**
Changes introduced in this pull request:
- Added a `chain_sync::consensus::Consensus` trait which currently has a single `validate_block` method for consensus specific rules. It also has an associated error type.
- Added a `blockchain/fil_consensus` crate with `FilecoinConsensus` implementing the `Consensus` trait. Moved the errors and validations specific to PoSt, tickets, beacons, DRAND, etc. into this crate. `FilecoinConsensus` also captures the generic beacon type and the `ProofValidator`. 
- The structs in `tipset_syncer` and `ChainMuxer` get a generic `consensus: C where C: Consensus` argument so they can delegate the custom validations to it. 
- They keep the common validation which is related to the messages and the ledger (gas, account sequences, state, fees). 
- Because the `ProofValidator` was moved into `FilecoinConsensus` I was able to remove it from all but one method in the RPC and `StateManager` where a block is assembled. 

**Other information and links**
This is laying some groundworks towards implementing [Hierarchical Consensus](https://github.com/protocol/ConsensusLab/blob/hc/spec/specs/hierarchical_consensus.md)

In the future we want to expand this interface to be able to generate blocks on a schedule, adhering to some hitherto unknown consensus protocol.

### CODE REVIEW

**Comment by lemmih** [comment]
Looks good. Mind adding license headers to the two new files? It can be done by running `make license` in the repository root.

**Comment by aakoshh** [comment]
Thanks @lemmih I added the license and resolved the clippy issues, also signed the CLA.

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 1]
```diff
@@ -1297,43 +1227,33 @@ async fn validate_block<
...
-                    TipsetRangeSyncerError::Validation(format!("Could not compute base fee: {}", e))
-                },
-            )?;
+        let base_fee = chain::compute_base_fee(
+            v_block_store.as_ref(),
+            &v_base_tipset,
+            smoke_height,
+        )
+        .map_err(|e| {
+            TipsetRangeSyncerError::<C>::Validation(format!("Could not compute base fee: {}", e))
```
Comment (LesnyRumcajs) [inline blockchain/chain_sync/src/tipset_syncer.rs:1252]:
```suggestion
            TipsetRangeSyncerError::<C>::Validation(format!("Could not compute base fee: {e}"))
```

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 2]
```diff
@@ -1398,159 +1320,53 @@ async fn validate_block<
...
+    validations.push(task::spawn(async move {
+        consensus
+            .validate_block(state_manager, v_block)
+            .map_err(|errs| {
+                // NOTE: Concatentating errors here means the wrapper type of error
+                // never surfaces, yet we always pay the cost of the generic argument.
+                // But there's no reason `validate_block` couldn't return a list of all
+                // errors instead of a single one that has all the error messages,
+                // removing the caller's ability to distinguish between them.
+                let mut msgs = vec![];
```
Comment (LesnyRumcajs) [inline blockchain/chain_sync/src/tipset_syncer.rs:1340]: does it make sense to `reserve` `errs.len()` upfront here?

**Review by LesnyRumcajs — COMMENTED** [review]
Really nice PR, a lot of documentation, thanks! Control question, did you check if it syncs on mainnet and calibnet?

File: blockchain/chain_sync/src/tipset_syncer.rs (hunk 2 above)
Comment (aakoshh) [reply blockchain/chain_sync/src/tipset_syncer.rs]: Thanks for mentioning this. I just noticed that not 10 lines further down I'm using map and collect instead of looping, so it wasn't very consistent. I moved both into a helper function where `with_capacity` is used.

**Comment by aakoshh** [comment]
> did you check if it syncs on mainnet and calibnet?

No. I see in the readme how to start it with calibnet, but not where I should get a `snapshot.car` file from.

How do you run these tests?

File: blockchain/chain_sync/src/tipset_syncer.rs (hunk 1 above)
Comment (LesnyRumcajs) [reply blockchain/chain_sync/src/tipset_syncer.rs:1264]: It's a new addition to Rust (since 1.58). https://www.rustnote.com/blog/format_strings.html

Definitely not a big deal though. :)

**Comment by LesnyRumcajs** [comment]
> > did you check if it syncs on mainnet and calibnet?
> 
> No. I see in the readme how to start it with calibnet, but not where I should get a `snapshot.car` file from.
> 
> How do you run these tests?

Right now those snapshots are generated with Lotus, there is a PR in progress to get out of this dependency. For now you can download a recent one [here](https://fra1.digitaloceanspaces.com/forest-snapshots/calibnet/calibnet-2022-07-01.car?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=3JPB6WZESYS26MFHEJW5%2F20220714%2Ffra1%2Fs3%2Faws4_request&X-Amz-Date=20220714T155556Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=e554ffe1b5c810fff9125e29647a538bfeda0bb5fcdc7f09d1511f868c0a7ed4)

To run calibnet on NV16 [this](https://github.com/ChainSafe/forest/pull/1642) PR needs to be merged first. Mainnet should be fine though.

Basically for mainnet:
```
 target/release/forest --target-peer-count 50 --encrypt-keystore false --import-snapshot ~/prj/snapshots/minimal_finality_stateroots_1974240_2022-07-11_10-00-00.car
```

Definitely, a piece of documentation we should work on!

**Comment by aakoshh** [comment]
@LesnyRumcajs thanks I will try. 

Meanwhile I just tried calibnet without any snapshot and it started syncing:
image

File: blockchain/chain_sync/src/tipset_syncer.rs (hunk 2 above)
Comment (aakoshh) [reply blockchain/chain_sync/src/tipset_syncer.rs]: I agree on the one liners - I went back to the collect method, although technically it's 4 lines because of the formatter :man_shrugging:

**Comment by aakoshh** [comment]
I found a link https://fil-chain-snapshots-fallback.s3.amazonaws.com/mainnet/minimal_finality_stateroots_latest.car to the mainnet snapshot but it's over 70Gb. I'm not going to download that, but tried the sync without it and it seems to work all right:

```console
❯ target/release/forest --target-peer-count 50 --encrypt-keystore false                                                                                                                                          ─╯
 2022-07-14T19:24:36.043Z WARN  forest::cli > No configurations found, using defaults.
 2022-07-14T19:24:36.044Z INFO  forest::daemon > Starting Forest daemon, version v0.2.2/unstable/65e3dcff
 2022-07-14T19:24:36.044Z INFO  forest_libp2p::service > Recovered libp2p keypair from "/home/aakoshh/.local/share/forest/libp2p/keypair"
 2022-07-14T19:24:36.044Z WARN  forest::daemon         > Warning: Keystore encryption disabled!
 2022-07-14T19:24:36.044Z WARN  key_management::keystore > Keystore does not exist, initializing new keystore at: "/home/aakoshh/.local/share/forest/keystore.json"
 2022-07-14T19:24:36.044Z INFO  utils                    > Permissions set to 0600 on File { fd: 6, path: "/home/aakoshh/.local/share/forest/keystore.json", read: false, write: true }
 2022-07-14T19:24:36.044Z INFO  forest::daemon           > Admin token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.TtyHyhQfcu5nhVPL2izQVhNgrW3MwNkBxWfOBqtPJp0
 2022-07-14T19:24:36.044Z INFO  metrics                  > Prometheus server started at 127.0.0.1:6116
 2022-07-14T19:24:36.090Z WARN  chain::store::chain_store > No previous chain state found
 2022-07-14T19:24:36.110Z INFO  genesis                   > Initialized genesis: BlockHeader: Cid(bafy2bzacecnamqgqmifpluoeldx7zzglxcljo6oja4vrmtj7432rphldpdmm2)
 2022-07-14T19:24:36.111Z INFO  forest::daemon            > Using network :: cannot get name
 2022-07-14T19:24:36.563Z INFO  forest::daemon            > JSON-RPC endpoint started at 127.0.0.1:1234
 2022-07-14T19:24:36.563Z INFO  rpc                       > Ready for RPC connections
 2022-07-14T19:24:37.974Z INFO  chain_sync::chain_muxer   > Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 0 -> NETWORK_HEAD = 1984009
Downloading headers 3614 / 1984009 [>-------------------------------------------------------------------------------------------------------------------------------------------------------] 0.18 % 147.05/s 224m ^C 2022-07-14T19:25:02.716Z WARN  forest::cli               > Got interrupt, shutting down...
 2022-07-14T19:25:02.717Z INFO  utils                     > Permissions set to 0600 on File { fd: 24, path: "/home/aakoshh/.local/share/forest/keystore.json", read: false, write: true }
 2022-07-14T19:25:02.746Z INFO  forest::daemon            > Forest finish shutdown.

```

**Comment by aakoshh** [comment]
@LesnyRumcajs the sync with the mainnet ran into some issue with 90 minutes left on the clock, it found no peers to request headers from. 

So I created a separate branch from this where I merged the PR you mentioned and tried syncing with calibnet instead, using the snapshot you linked to. That seems to have worked:

```console
 ./target/release/forest --chain calibnet --import-snapshot ../calibnet-2022-07-01.car --target-peer-count 50 --encrypt-keystore false                                                                          ─╯
 2022-07-14T22:04:08.309Z WARN  forest::cli > No configurations found, using defaults.
 2022-07-14T22:04:08.309Z INFO  forest::daemon > Starting Forest daemon, version v0.2.2/unstable/e6206af3
 2022-07-14T22:04:08.309Z INFO  forest_libp2p::service > Recovered libp2p keypair from "/home/aakoshh/.local/share/forest/libp2p/keypair"
 2022-07-14T22:04:08.309Z WARN  forest::daemon         > Warning: Keystore encryption disabled!
 2022-07-14T22:04:08.310Z INFO  forest::daemon         > Admin token: eyJ0eXAiOiJKV1QiLCJhbGciOi...
 2022-07-14T22:04:08.310Z INFO  metrics                > Prometheus server started at 127.0.0.1:6116
 2022-07-14T22:04:08.376Z INFO  genesis                > Initialized genesis: BlockHeader: Cid(bafy2bzacecz3trtejxtzix4f4eebs7dekm6snfsmvffiqz2rfx7iwgsgtieq4)
 2022-07-14T22:04:08.376Z INFO  forest::daemon         > Using network :: cannot get name
 2022-07-14T22:04:08.376Z INFO  genesis                > Importing chain from snapshot
 2022-07-14T22:04:08.376Z INFO  genesis                > Reading file...
Scanning blockchain 1082530 / 1087290 [=====================================================================================================================================================] 99.56 % 8530.18/s 1s  2022-07-14T22:06:45.956Z INFO  genesis                > Accepting [Cid(bafy2bzacea7c2aqujxac37aeizzg5hr2eudwumppgcek5qvxmrkp3gjv7jlm6), Cid(bafy2bzacebl2enajghfcgzvgxf66syddba4adfsexz74jca3ygxbdb2jae7bg), Cid(bafy2bzacedjwvigkrlyzvn6qvxzygw7u2qqxqywc7i7wnwqkcxfk7rf75lovw), Cid(bafy2bzacecrx6cmjyobmuaexgascuqh3ad4kxh3ff2bdxjecmusvwvs22w65w), Cid(bafy2bzacebrv2rkkqvqxuoeou2a55xwhjqsyyujic54sfsfzxuqjo2tirfk4a), Cid(bafy2bzacedd76nyajhq2bphw3aztdmtlwtrrawcvvkrcebdq6lreabqig3wyy)] as new head.
 2022-07-14T22:06:46.433Z INFO  forest::daemon         > JSON-RPC endpoint started at 127.0.0.1:1234
 2022-07-14T22:06:46.433Z INFO  rpc                    > Ready for RPC connections
 2022-07-14T22:06:48.521Z INFO  chain_sync::chain_muxer > Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 1087290 -> NETWORK_HEAD = 1125853
Downloading headers 38563 / 38563 [=============================================================================================================================================================] 100.00 % 97.03/s  2022-07-14T22:13:27.051Z INFO  chain_sync::tipset_syncer > Validating tipset: EPOCH = 1087290
 2022-07-14T22:13:30.707Z INFO  chain_sync::tipset_syncer > Validating tipset: EPOCH = 1087291
 2022-07-14T22:13:30.764Z INFO  chain_sync::tipset_syncer > Validating tipset: EPOCH = 1087292
 2022-07-14T22:13:31.161Z INFO  chain_sync::tipset_syncer > Validating tipset: EPOCH = 1087293
...
```

**Comment by LesnyRumcajs** [comment]
@aakoshh You can't really sync without the snapshots as Forest doesn't have any migrations at the moment - only pure NV16. Even with those in place, it would take several weeks to sync. But it's a good point for exiting early instead of raising hopes and expectations!

Glad it works, thanks for checking it out!

**Review by LesnyRumcajs — APPROVED** [review]

File: blockchain/chain_sync/src/consensus.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+/// The `Consensus` trait encapsulates consensus specific rules of validation
+/// and block creation. Behind the scenes they can farm out the total ordering
+/// of transactions to an arbitrary consensus engine, but in the end they
+/// package the transactions into Filecoin compatible blocks.
+///
+/// Not all fields will be made use of, however, so the validation of these
+/// blocks at least partially have to be trusted to the `Consensus` component.
+///
+/// Common rules for message ordering will be followed, and can be validated
+/// outside by the host system during chain syncrhonization.
```
Comment (elmattic) [inline blockchain/chain_sync/src/consensus.rs:26]: letter swap in synchronization

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 4]
```diff
@@ -53,41 +50,33 @@ use state_tree::StateTree;
...
     AlreadySynced,
 }
 
 #[derive(Debug, Error)]
-pub enum TipsetRangeSyncerError {
-    #[error("Tipset added to range syncer does share the same epoch and parents")]
-    InvalidTipsetAdded,
+pub enum TipsetRangeSyncerError<C: Consensus> {
+    // #[error("Tipset added to range syncer does share the same epoch and parents")]
+    // InvalidTipsetAdded,
```
Comment (elmattic) [inline blockchain/chain_sync/src/tipset_syncer.rs:65]: Do we really want to keep this variant commented?

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 5]
```diff
@@ -142,10 +111,23 @@ pub enum TipsetRangeSyncerError {
...
-    #[error("[INSECURE-POST-VALIDATION] {0}")]
-    InsecurePostValidation(String),
     #[error("Loading tipset parent from the store failed: {0}")]
     TipsetParentNotFound(ChainStoreError),
+    #[error("Consensus error: {0}")]
+    ConsensusError(C::Error),
+}
+
+impl<C: Consensus> TipsetRangeSyncerError<C> {
+    /// Concatanate all validation error messages into one comma separated version.
```
Comment (elmattic) [inline blockchain/chain_sync/src/tipset_syncer.rs:121]: typo in doc comment on concatenate

**Review by elmattic — CHANGES_REQUESTED** [review]
lgtm, nice refactor!

File: blockchain/chain_sync/src/tipset_syncer.rs (hunk 4 above)
Comment (aakoshh) [reply blockchain/chain_sync/src/tipset_syncer.rs:65]: I don't know, I just saw it wasn't used and tried commenting out to make sure. Wanted to check the other ones as well but there were too many; I'll have another go. If you don't think it's needed I'll remove it now.

**Review by elmattic — APPROVED** [review]
